### PR TITLE
Restricted Pairwise union reach surface model

### PR DESCRIPTION
**Pull request by uakyol**

### CODE REVIEW

File: src/models/pairwise_union_reach_surface.py [hunk 1]
```diff
@@ -22,115 +23,229 @@
...
+
+    def get_inequality_constraints(self):
+        """Returns the inequality constraint for the Quadratic Programming solver.
+
+        3 types of inequality constraints are stacked in matrix G and vector h:
+
+          G1: a_i_j >= 0 for all i,j -> all parameters are non negative
+          G2: sum_{j=1..p} a_i_j <= 1 for all i -> row sums are no more than one
+          G3: sum_{i=1..p} a_i_j <= 1 for all j -> col sums are no more than one
+        Returns:
```
Comment (jiayu-google) [inline src/models/pairwise_union_reach_surface.py:120]: An empty line before Returns

File: src/models/restricted_pairwise_union_reach_surface.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+from cvxopt import solvers
+from typing import Iterable
+from wfa_planning_evaluation_framework.models.reach_point import ReachPoint
+from wfa_planning_evaluation_framework.models.reach_curve import ReachCurve
+from wfa_planning_evaluation_framework.models.pairwise_union_reach_surface import (
+    PairwiseUnionReachSurface,
+)
+
+
+class RestrictedPairwiseUnionReachSurface(PairwiseUnionReachSurface):
```
Comment (jiayu-google) [inline src/models/restricted_pairwise_union_reach_surface.py:29]: Good design to treat RestrictedPairwise as a subclass of Pairwise!

**Review by jiayu-google — COMMENTED** [review]
Ugur, great job! The PR looks good to me so far. 

The PR consists of several components. I'm wondering if there is an EASY way to split the PR into 2. If not, I'll continue reviewing it tomorrow.

File: src/models/restricted_pairwise_union_reach_surface.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    def construct_a_from_lambda(self, lbd):
+        """Get value of flattened a matrix from lamdas.
+
+        Args:
+          lbd: a length p vector indicating lambda_i for each pub.
+
+        Returns:
+          the value of flattened a matrix.
+        """
+        self._a = np.ones(self._p * self._p)
```
Comment (matthewclegg) [inline src/models/restricted_pairwise_union_reach_surface.py:72]: I suppose this function could be written as:

p = len(lbd)
A = lbd.reshape(p,1) * lbd.reshape(1,p) - (np.eye(p) * lbd)**2
return A.reshape(,1)

File: src/models/restricted_pairwise_union_reach_surface.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                    )
+                    * 2
+                )
+                for i in range(self._p - 1)
+                for j in range(i, self._p)
+            ]
+        )
+        return reach_sum - overlap
+
+    def loss(self, lbd):
```
Comment (matthewclegg) [inline src/models/restricted_pairwise_union_reach_surface.py:119]: _loss
Also, add typing information.

File: src/models/restricted_pairwise_union_reach_surface.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+        cons = []
+        for i in range(self._p):
+            # All lambdas are non negative : lbd[i] >= 0
+            cons.append({"type": "ineq", "fun": lambda x: x[i]})
+            # Lambda j sum times Lambda i is less than 1 : 1 - lbd[i] * sum(lbd) >= 0
+            cons.append({"type": "ineq", "fun": lambda x: 1 - x[i] * sum(x)})
+        return cons
+
+    def evaluate_point(self, lbd, reach_vector):
```
Comment (matthewclegg) [inline src/models/restricted_pairwise_union_reach_surface.py:92]: _evaluate_point
Also, add typing information.

File: src/models/restricted_pairwise_union_reach_surface.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+        Returns:
+          the value of flattened a matrix.
+        """
+        self._a = np.ones(self._p * self._p)
+        for i in range(self._p):
+            for j in range(self._p):
+                self._a[i * self._p + j] = lbd[i] * lbd[j] if i != j else 0
+
+    def get_constraints(self):
```
Comment (matthewclegg) [inline src/models/restricted_pairwise_union_reach_surface.py:77]: _get_constraints
Also, add typing information.

File: src/models/restricted_pairwise_union_reach_surface.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        res = minimize(
+            fun=lambda x: self.loss(x),
+            x0=np.array([2 / self._p] * self._p) / 2,
+            constraints=cons,
+            method="trust-constr",
+            options={"disp": True},
+        )
+        self.construct_a_from_lambda(res["x"])
+
+    def construct_a_from_lambda(self, lbd):
```
Comment (matthewclegg) [inline src/models/restricted_pairwise_union_reach_surface.py:63]: _construct_a_from_lambda
Also, add typing information.

**Review by matthewclegg — CHANGES_REQUESTED** [review]

File: src/models/tests/restricted_pairwise_union_reach_surface_test.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        max_reaches = [
+            universe_size * (decay_rate ** pub_num) for pub_num in range(num_publishers)
+        ]
+        return [
+            LinearCappedReachCurve([ReachPoint([max_reach], (max_reach,))])
+            for max_reach in max_reaches
+        ]
+
+    def generate_sample_matrix_a(self, num_publishers):
+        lbd = np.array([[2 / num_publishers] for _ in range(num_publishers)])
```
Comment (jiayu-google) [inline src/models/tests/restricted_pairwise_union_reach_surface_test.py:81]: Following what Matthew did before: for testing we may set lbd = 0, i.e., non-overlapping reach surface. The model should have a perfect prediction for this. Would you like to try it?

**Review by jiayu-google — APPROVED** [review]
Thank you Ugur!

File: src/models/tests/restricted_pairwise_union_reach_surface_test.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+            impressions = [
+                random_generator.uniform(0, universe_size / 2)
+                for _ in range(len(reach_curves))
+            ]
+            reach_points.append(
+                self.generate_true_reach(true_a, reach_curves, impressions)
+            )
+        return reach_points
+
+    def test_by_impressions(self):
```
[the remark this answers is not in the document]
Comment (jiayu-google) [reply src/models/tests/restricted_pairwise_union_reach_surface_test.py:104]: Perhaps try `cobyla`. We do need to carefully investigate the corner cases when the scipy minimizer is used here. Perhaps can do it in Q3.

File: src/models/restricted_pairwise_union_reach_surface.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+from wfa_planning_evaluation_framework.models.reach_curve import ReachCurve
+from wfa_planning_evaluation_framework.models.pairwise_union_reach_surface import (
+    PairwiseUnionReachSurface,
+)
+
+
+class RestrictedPairwiseUnionReachSurface(PairwiseUnionReachSurface):
+    """Models reach with the pairwise union overlap model."""
+
+    def __init__(
```
[the remark this answers is not in the document]
Comment (uakyol) [reply src/models/restricted_pairwise_union_reach_surface.py:32]: thanks for catching

File: src/models/restricted_pairwise_union_reach_surface.py [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+                for reach_point in self._data
+            ]
+        )
+        cons = self.get_constraints()
+        res = minimize(
+            fun=lambda x: self.loss(x),
+            x0=np.array([2 / self._p] * self._p) / 2,
+            constraints=cons,
+            method="trust-constr",
+            options={"disp": True},
```
[the remark this answers is not in the document]
Comment (uakyol) [reply src/models/restricted_pairwise_union_reach_surface.py:59]: thanks

File: src/models/restricted_pairwise_union_reach_surface.py (hunk 3 above)
Comment (uakyol) [reply src/models/restricted_pairwise_union_reach_surface.py:72]: thanks

File: src/models/tests/restricted_pairwise_union_reach_surface_test.py (hunk 8 above)
Comment (uakyol) [reply src/models/tests/restricted_pairwise_union_reach_surface_test.py:81]: done

File: src/models/tests/restricted_pairwise_union_reach_surface_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (uakyol) [reply src/models/tests/restricted_pairwise_union_reach_surface_test.py:104]: I don't think this is an error, rather a warning. Looking into it.

[the remark this answers is not in the document]
Comment (AaronRanAn) [reply src/models/tests/restricted_pairwise_union_reach_surface_test.py:104]: Started to look into this PR. @uakyol has this error/warning being resolved?

[the remark this answers is not in the document]
Comment (uakyol) [reply src/models/tests/restricted_pairwise_union_reach_surface_test.py:104]: Sorry for the late reply Aaron, I am investigating this and thought would resolve sooner. It seems I'll take another day. This is warning itself is not a problem but we are making sure that there aren't any other unforeseen issues this creates by getting some ad hoc runs

[the remark this answers is not in the document]
Comment (AaronRanAn) [reply src/models/tests/restricted_pairwise_union_reach_surface_test.py:104]: Got it, thanks for the update! Let's unblock this first and resolve the warning later if it raises.

**Review by AaronRanAn — APPROVED** [review]
LGTM, please merge to unblock and keep us updated on the warning resolution later.

**Comment by uakyol** [comment]
Thanks a lot @AaronRanAn . We just finished taking a run to make sure that this is not causing unforeseen problems. I also added code to handle the warnings as well as bad optimizer runs.

**Review by matthewclegg — APPROVED** [review]
Thanks, Ugur! Great work!
